From c7e06b9013bf0b47eef9037f3e3230579162838f Mon Sep 17 00:00:00 2001
From: AbdAlRahman Gad <abdobngad@gmail.com>
Date: Tue, 08 Oct 2024 02:43:31 +0000
Subject: [PATCH] Fix dangling pointers by nullifying child view raw pointers before removal

Signed-off-by: AbdAlRahman Gad <abdobngad@gmail.com>
Bug: 366010379
Change-Id: I98b2b7582af45d2050f63f9429f1143564825edd
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5861129
Commit-Queue: Peter Kasting <pkasting@chromium.org>
Reviewed-by: Charles Meng <charlesmeng@chromium.org>
Reviewed-by: Peter Kasting <pkasting@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1365290}
---

--- a/AUTHORS
+++ b/AUTHORS
@@ -17,6 +17,7 @@
 Aaron Leventhal <aaronlevbugs@gmail.com>
 Aaron Randolph <aaron.randolph@gmail.com>
 Aaryaman Vasishta <jem456.vasishta@gmail.com>
+AbdAlRahman Gad <abdobngad@gmail.com>
 Abdu Ameen <abdu.ameen000@gmail.com>
 Abdullah Abu Tasneem <a.tasneem@samsung.com>
 Abhijeet Kandalkar <abhijeet.k@samsung.com>
--- a/chrome/browser/ui/views/frame/browser_view.cc
+++ b/chrome/browser/ui/views/frame/browser_view.cc
@@ -1125,9 +1125,30 @@
     global_registry->set_registry_for_active_window(nullptr);
   }
 
-  // `watermark_view_` is a raw pointer to a child view, so it needs to be set
-  // to null before `RemoveAllChildViews()` is called to avoid dangling.
+  // These are raw pointers to child views, so they need to be set to null
+  // before `RemoveAllChildViews()` is called to avoid dangling.
+  frame_ = nullptr;
+  top_container_ = nullptr;
+  web_app_frame_toolbar_ = nullptr;
+  web_app_window_title_ = nullptr;
+  tab_strip_region_view_ = nullptr;
+  tabstrip_ = nullptr;
+  webui_tab_strip_ = nullptr;
+  toolbar_ = nullptr;
+  contents_separator_ = nullptr;
+  loading_bar_ = nullptr;
+  find_bar_host_view_ = nullptr;
+  download_shelf_ = nullptr;
+  infobar_container_ = nullptr;
+  contents_web_view_ = nullptr;
+  devtools_web_view_ = nullptr;
   watermark_view_ = nullptr;
+  contents_container_ = nullptr;
+  unified_side_panel_ = nullptr;
+  right_aligned_side_panel_separator_ = nullptr;
+  left_aligned_side_panel_separator_ = nullptr;
+  side_panel_rounded_corner_ = nullptr;
+  toolbar_button_provider_ = nullptr;
 
   // Child views maintain PrefMember attributes that point to
   // OffTheRecordProfile's PrefService which gets deleted by ~Browser.
@@ -4720,12 +4741,10 @@
     }
   } else if (webui_tab_strip_) {
     top_container_->RemoveChildView(webui_tab_strip_);
-    delete webui_tab_strip_;
-    webui_tab_strip_ = nullptr;
+    webui_tab_strip_.ClearAndDelete();
 
     top_container_->RemoveChildView(loading_bar_);
-    delete loading_bar_;
-    loading_bar_ = nullptr;
+    loading_bar_.ClearAndDelete();
   }
   GetBrowserViewLayout()->set_webui_tab_strip(webui_tab_strip_);
   GetBrowserViewLayout()->set_loading_bar(loading_bar_);
--- a/chrome/browser/ui/views/frame/browser_view.h
+++ b/chrome/browser/ui/views/frame/browser_view.h
@@ -1058,7 +1058,7 @@
   void UpdateFullscreenAllowedFromPolicy(bool allowed_without_policy);
 
   // The BrowserFrame that hosts this view.
-  raw_ptr<BrowserFrame, DanglingUntriaged> frame_ = nullptr;
+  raw_ptr<BrowserFrame> frame_ = nullptr;
 
   // The Browser object we are associated with.
   std::unique_ptr<Browser> browser_;
@@ -1092,29 +1092,25 @@
   // The view that manages the tab strip, toolbar, and sometimes the bookmark
   // bar. Stacked top in the view hiearachy so it can be used to slide out
   // the top views in immersive fullscreen.
-  raw_ptr<TopContainerView, AcrossTasksDanglingUntriaged> top_container_ =
-      nullptr;
+  raw_ptr<TopContainerView> top_container_ = nullptr;
 
   // Menu button and page status icons. Only used by web-app windows.
-  raw_ptr<WebAppFrameToolbarView, DanglingUntriaged> web_app_frame_toolbar_ =
-      nullptr;
+  raw_ptr<WebAppFrameToolbarView> web_app_frame_toolbar_ = nullptr;
 
   // Normally the BrowserNonClientFrameView is responsible for rendering the
   // title of a window when appropriate. However for web applications the title
   // needs to be more integrated with other UI components part of BrowserView,
   // so have a title Label for them here.
-  raw_ptr<views::Label, DanglingUntriaged> web_app_window_title_ = nullptr;
+  raw_ptr<views::Label> web_app_window_title_ = nullptr;
 
   // The view that contains the tabstrip, new tab button, and grab handle space.
-  raw_ptr<TabStripRegionView, AcrossTasksDanglingUntriaged>
-      tab_strip_region_view_ = nullptr;
+  raw_ptr<TabStripRegionView> tab_strip_region_view_ = nullptr;
 
   // The TabStrip.
-  raw_ptr<TabStrip, AcrossTasksDanglingUntriaged> tabstrip_ = nullptr;
+  raw_ptr<TabStrip> tabstrip_ = nullptr;
 
   // the webui based tabstrip, when applicable. see https://crbug.com/989131.
-  raw_ptr<WebUITabStripContainerView, DanglingUntriaged> webui_tab_strip_ =
-      nullptr;
+  raw_ptr<WebUITabStripContainerView> webui_tab_strip_ = nullptr;
 
   // Allows us to react to changes in accessibility mode.
   // TODO(dfried): this is only used to disable WebUI tabstrip (see above) while
@@ -1124,7 +1120,7 @@
   std::unique_ptr<AccessibilityModeObserver> accessibility_mode_observer_;
 
   // The Toolbar containing the navigation buttons, menus and the address bar.
-  raw_ptr<ToolbarView, AcrossTasksDanglingUntriaged> toolbar_ = nullptr;
+  raw_ptr<ToolbarView> toolbar_ = nullptr;
 
   // The OverlayView for the widget, which is used to host `top_container_`
   // during immersive reveal.
@@ -1154,57 +1150,47 @@
   std::unique_ptr<BookmarkBarView> bookmark_bar_view_;
 
   // Separator between top container and contents.
-  raw_ptr<views::View, AcrossTasksDanglingUntriaged> contents_separator_ =
-      nullptr;
+  raw_ptr<views::View> contents_separator_ = nullptr;
 
   // Loading bar (part of top container for / WebUI tab strip).
-  raw_ptr<TopContainerLoadingBar, DanglingUntriaged> loading_bar_ = nullptr;
+  raw_ptr<TopContainerLoadingBar> loading_bar_ = nullptr;
 
   // The do-nothing view which controls the z-order of the find bar widget
   // relative to views which paint into layers and views with an associated
   // NativeView.
-  raw_ptr<View, AcrossTasksDanglingUntriaged> find_bar_host_view_ = nullptr;
+  raw_ptr<View> find_bar_host_view_ = nullptr;
 
   // The download shelf.
-  raw_ptr<DownloadShelf, DanglingUntriaged> download_shelf_ = nullptr;
+  raw_ptr<DownloadShelf> download_shelf_ = nullptr;
 
   // The InfoBarContainerView that contains InfoBars for the current tab.
-  raw_ptr<InfoBarContainerView, AcrossTasksDanglingUntriaged>
-      infobar_container_ = nullptr;
+  raw_ptr<InfoBarContainerView> infobar_container_ = nullptr;
 
   // The view that contains the selected WebContents.
-  raw_ptr<ContentsWebView, AcrossTasksDanglingUntriaged> contents_web_view_ =
-      nullptr;
+  raw_ptr<ContentsWebView> contents_web_view_ = nullptr;
 
   // The view that contains devtools window for the selected WebContents.
-  raw_ptr<views::WebView, AcrossTasksDanglingUntriaged> devtools_web_view_ =
-      nullptr;
+  raw_ptr<views::WebView> devtools_web_view_ = nullptr;
 
   // The view that overlays a watermark on the contents container.
   raw_ptr<enterprise_watermark::WatermarkView> watermark_view_ = nullptr;
 
   // The view managing the devtools and contents positions.
   // Handled by ContentsLayoutManager.
-  raw_ptr<views::View, AcrossTasksDanglingUntriaged> contents_container_ =
-      nullptr;
+  raw_ptr<views::View> contents_container_ = nullptr;
 
   // The side panel aligned to the left or the right side of the browser window
   // depending on the kSidePanelHorizontalAlignment pref's value.
   // Conceptually this member should exist if and only if the
   // side_panel_coordinator is created.
-  raw_ptr<SidePanel, AcrossTasksDanglingUntriaged> unified_side_panel_ =
-      nullptr;
-  raw_ptr<views::View, AcrossTasksDanglingUntriaged>
-      right_aligned_side_panel_separator_ = nullptr;
-  raw_ptr<views::View, AcrossTasksDanglingUntriaged>
-      left_aligned_side_panel_separator_ = nullptr;
-  raw_ptr<views::View, AcrossTasksDanglingUntriaged>
-      side_panel_rounded_corner_ = nullptr;
+  raw_ptr<SidePanel> unified_side_panel_ = nullptr;
+  raw_ptr<views::View> right_aligned_side_panel_separator_ = nullptr;
+  raw_ptr<views::View> left_aligned_side_panel_separator_ = nullptr;
+  raw_ptr<views::View> side_panel_rounded_corner_ = nullptr;
 
   // Provides access to the toolbar buttons this browser view uses. Buttons may
   // appear in a hosted app frame or in a tabbed UI toolbar.
-  raw_ptr<ToolbarButtonProvider, AcrossTasksDanglingUntriaged>
-      toolbar_button_provider_ = nullptr;
+  raw_ptr<ToolbarButtonProvider> toolbar_button_provider_ = nullptr;
 
   // The handler responsible for showing autofill bubbles.
   std::unique_ptr<autofill::AutofillBubbleHandler> autofill_bubble_handler_;
